src/Processors/QueryPlan/JoinStepLogical.cpp

```diff
@@ -781,4 +785,100 @@ std::optional<ActionsDAG> JoinStepLogical::getFilterActions(JoinTableSide side,
     return {};
 }
 
+void JoinStepLogical::serializeSettings(QueryPlanSerializationSettings & settings) const
+{
+    join_settings.updatePlanSettings(settings);
+    sorting_settings.updatePlanSettings(settings);
+}
+
+void JoinStepLogical::serialize(Serialization & ctx) const
+{
+    if (prepared_join_storage)
+        throw Exception(ErrorCodes::NOT_IMPLEMENTED, "Serialization of JoinStepLogical with prepared storage is not implemented");
+
+    if (swap_inputs)
+        throw Exception(ErrorCodes::NOT_IMPLEMENTED, "Serialization of JoinStepLogical with swapped inputs is not implemented");
+
+    UInt8 flags = 0;
+    if (use_nulls)
+        flags |= 1;
+
+    writeIntBinary(flags, ctx.out);
+
+    JoinActionRef::ActionsDAGRawPtrs dags = {
+        expression_actions.left_pre_join_actions.get(),
+        expression_actions.right_pre_join_actions.get(),
+        expression_actions.post_join_actions.get()};
+
+    writeVarUInt(dags.size(), ctx.out);
+    for (const auto * dag : dags)
+        dag->serialize(ctx.out, ctx.registry);
```

Comment:
When we generalize a step for multiple inputs, there may be ambiguity in matching inputs and the DAGs.
In my current local draft, I store DAGs in a map where the key is a 64-bit mask representing the inputs, and the values are the corresponding DAGs. We could also serialize the index of the input corresponding to each DAG (1 for the left input, 2 for the right input, and 3 for “post,” indicating that it depends on both the first and second inputs).

Reply:
Ok. We can do this (need to add +1 to the index :)
Also, I don't worry about breaking the serialization if needed. We can worry about compatibility later.